src/org/javarosa/core/model/ItemsetBinding.java

```diff
@@ -183,4 +198,54 @@ public void writeExternal(DataOutputStream out) throws IOException {
         ExtUtil.writeBool(out, copyMode);
     }
 
+    public TreeReference getNodesetRef() {
+        return nodeset
+            .map(ns -> {
+                TreeReference refAt = ns.getRefAt(0);
+                refAt.setInstanceName(ns.instance.getName());
+                return refAt;
+            })
+            .orElseGet(() -> nodesetRef.orElse(null));
+    }
+
+    public TreeReference getLabelRef() {
+        return nodeset
+            .map(ns -> computeRef(ns.getRefAt(0), labelExpr).get())
```

Comment:
Am I understanding correctly that this will just get the reference for the first node in the nodeset which could be completely different than the ones for other nodes in the set but that it doesn't matter because it's just used for surface error checking?

Reply:
It looks like it, but I couldn't be sure. I'm thinking that moving all the validation and coherence checks inside ItemsetBinding to make sure about that.